File: internal/factory/dlq_recovery.go
package factory

import "slices"

// dlqRecoveryCapableSinkTypes are the sink types that can also be read back as
// a source, which is what "Prioritize DLQ on startup" and the Drain DLQ button
// both need: the engine wraps the dead-letter sink in a PrioritySource and
// reads parked messages out of it again.
//
// The editor used to keep its own copy of this list. It had drifted in both
// directions: four types it advertised as recovery-capable were not sources at
// all, so the checkbox was offered and StartWorkflow then refused the
// workflow; and nine types that are both sink and source were missing, so the
// editor greyed the checkbox out and the feature was simply unreachable for
// them. The list lives here now, beside the switches that decide it, and
// TestDLQRecoveryListMatchesFactory fails if either switch moves without it.
var dlqRecoveryCapableSinkTypes = []string{
	"cassandra",
	"clickhouse",
	"discord",
	"dynamics365",
	"eventstore",
	"facebook",
	"file",
	"ftp",
	"googlesheets",
	"http",
	"instagram",
	"kafka",
	"linkedin",
	"mariadb",
	"metis",
	"mongodb",
	"mqtt",
	"mssql",
	"mysql",
	"nats",
	"oracle",
	"postgres",
	"rabbitmq",
	"rabbitmq_queue",
	"redis",
	"s3",
	"sap",
	"slack",
	"sqlite",
	"tiktok",
	"twitter",
	"websocket",
	"yugabyte",
}

// DLQRecoveryCapableSinkTypes returns the sink types usable as a dead-letter
// sink that the engine can also drain. The result is a copy: callers sort and
// filter it.
func DLQRecoveryCapableSinkTypes() []string {
	return slices.Clone(dlqRecoveryCapableSinkTypes)
}

// SupportsDLQRecovery reports whether a dead-letter sink of this type can be
// read back for PrioritizeDLQ or an on-demand drain.
func SupportsDLQRecovery(sinkType string) bool {
	return slices.Contains(dlqRecoveryCapableSinkTypes, sinkType)
}

File: ui/src/utils/dlqRecovery.ts
// "Prioritize DLQ on startup" and the Drain DLQ button both read messages back
// out of the dead-letter sink, which only works for sink types Hermod can also
// open as a source.
//
// The editor used to answer this from a literal list of 25 sink types written
// into SidebarDrawer.tsx. It had drifted from the factory in both directions:
// four types it called recovery-capable are not sources at all, so the
// checkbox was offered and StartWorkflow then refused the workflow; and nine
// types that are both sink and source were missing, so the checkbox was
// disabled and the feature was unreachable for them. The list now comes from
// GET /api/sinks/capabilities/dlq-recovery, which the factory owns.

export interface DLQSinkLike {
  type?: string;
}

/**
 * Whether the editor should treat this dead-letter sink as drainable.
 *
 * `capableTypes` is null/undefined while the capability request is in flight.
 * That is reported as capable on purpose: flashing a warning and disabling the
 * checkbox for a moment on every page load is worse than being briefly
 * optimistic, and the engine validates the choice on start either way.
 */
export function dlqRecoverySupported(
  sink: DLQSinkLike | null | undefined,
  capableTypes: string[] | null | undefined,
): boolean {
  if (!sink) return true;
  if (!capableTypes) return true;
  return typeof sink.type === 'string' && capableTypes.includes(sink.type);
}
